Test (missing) RSA key support in Secure Session

iamnotacake opened the pull request:
Add few specific tests to Go, NodeJS, Python, Ruby, C++ wrappers. These tests make sure the wrapper API gives error (exception) when we try to create Secure Session with a key different from EC private one.

While adding these tests, few bugs were discovered in Python and Ruby wrappers. To be precise, it was wrong way of checking whether `secure_session_create()` returned `NULL`. The tests were not failing as expected. This PR also brings fixes for them.

Speaking about other wrappers, some of them seem to handle this kind of errors as well: [Java (JNI)](https://github.com/cossacklabs/themis/blob/master/jni/themis_session.c#L577-L587), [PHP](https://github.com/cossacklabs/themis/blob/master/src/wrappers/themis/php/php_themis.c#L107-L111), another [PHP](https://github.com/cossacklabs/themis/blob/master/src/wrappers/themis/php7/php_session.c#L98-L103), [JS (WASM)](https://github.com/cossacklabs/themis/blob/master/src/wrappers/themis/wasm/src/secure_session.js#L177-L185)
And others may be missing the check: [Objective C](https://github.com/cossacklabs/themis/blob/master/src/wrappers/themis/Obj-C/objcthemis/ssession.m#L35-L38)

I've also updated `.gitignore` to include some stuff like generated Ruby packages, some Python package metadata, another Rust `target` directory.

## Checklist

- [X] Change is covered by automated tests
- [X] The [coding guidelines] are followed
- [x] Changelog is updated

[coding guidelines]: https://github.com/cossacklabs/themis/blob/master/CONTRIBUTING.md

ilammy commented:
> And others may be missing the check: Objective C

Yep, it definitely is missing the check. The correct way will be to check for `self.session` value and return `nil` out of the initializer.

Please keep a note of which wrappers still need checks and tests.

ilammy commented on gothemis/session/session_test.go:
```diff
@@ -223,6 +223,29 @@ func testSession(keytype int, t *testing.T) {
 
+func testInvalidKeyHandling(t *testing.T) {
+	kpa, err := keys.New(keys.TypeRSA)
+	if nil != err {
+		t.Error(err)
+		return
+	}
```
Within Go conditions assignments are not allowed. Therefore, a customary form of error checks exists:

```go
if err != nil {
        // handle it
}
```

Also, `testing.T` has an alternate method for reporting fatal errors:

```go
t.Fatal("sum ting wong", err)
```

which immediately stops the test after reporting an issue.

iamnotacake replied:
> Are more wrappers expected in this PR of this is all for now?

No, that's all. But I would appreciate if you can add that little check you mentioned earlier to ObjC wrapper and test it on some Apple device (at least, compilation). And then maybe show `git format-patch` result so I can do `git am`. That shouldn't affect existing ObjC tests as well as all the code that actually uses this wrapper, I hope.

ilammy commented on gothemis/session/session_test.go:
```diff
@@ -223,6 +223,29 @@ func testSession(keytype int, t *testing.T) {
 	<-finCh
 }
 
+func testInvalidKeyHandling(t *testing.T) {
```
Since this test is not parameterized, it can be just standalone `TestInvalidKeyHandling`, without invoking it from TestSession.

ilammy commented on src/wrappers/themis/jsthemis/test/test.js:
```diff
@@ -236,6 +236,7 @@ describe("jsthemis", function(){
         let serverID = Buffer.from('server')
         let messageA = Buffer.from('testing, testing')
         let messageB = Buffer.from('everything seems to be in order')
+        let rsaKeyPair = addon.KeyPair(Buffer.from("UlJBMgAABJBm8NGvTdCr5R7/iNvQy+WrC6DsYPg5ze5dc1/UxsthqEE4t+4Euwq6IVdddjisxUBvzWQ0VuyQbaExbCrrIaMYHgF9DSNOyrCznrHKBlNZrfxM+pHIkVB6q+b2hLEX4j3wnMZKE0dE0W3SGO/p0oYCsF0gjC7kNTjsh67WwjsIo6EqOrAG7RF9p3UaGUNm4tGEKCAM4NE8d8URcaWSIhQIHeOrVVbziWdMFiOX2GRGwpi2lWFxTLhI1Tyyov+NLy/J5b7NL8n7qusC13eG2XPMrEucuhZtihczn5l4wGkrPi/+jAW+Kn09rQERAPe1rOXHcpu2PzVOdO0bBkq ...
```
I guess it wouldn't hurt to note in the comment that JsThemis does not support generating RSA key pairs, that's why a hardcoded sample value is used here.

ilammy commented on tests/pythemis/test_ssession.py:
```diff
@@ -133,6 +134,35 @@ def test_ssession2(self):
+        keypair1 = skeygen.GenerateKeyPair(skeygen.KEY_PAIR_TYPE.EC)
+        user_id1 = b'user_id1'
+
+        # public key should be rejected
+        self.assertRaises(exception.ThemisError, lambda: ssession.SSession(
+                user_id1, keypair1.export_public_key(), transport))
```
Exception assertions have a more convenient form:

```python
with self.assertRaises(exception.ThemisError, msg="public key should be rejected"):
    ssession.SSession(user_id1, keypair1.export_public_key(), transport)
```

ilammy commented on src/wrappers/themis/ruby/lib/rbthemis.rb:
```diff
@@ -274,7 +274,7 @@ def initialize(id, private_key, transport)
         id_buf, id_length, private_key_buf, private_key_length, @callbacks)
 
-      raise ThemisError, 'Secure Session failed creating' unless @session
+      raise ThemisError, 'Secure Session failed creating' unless !@session.null?
```
You can use `if` here instead of  `unless` to avoid extra negation:

```ruby
raise ThemisError, 'Secure Session failed creating' if @session.null?
```

which makes the code a bit more readable.

ilammy commented on tests/rbthemis/ssession_test.rb:
```diff
@@ -93,4 +94,34 @@ def test_session
     ].each(&:join)
   end
+
+  def test_invalid_key_public_ec
+    begin
+      session = Themis::Ssession.new(
+        'client', @client_pub_key, CallbacksForThemis.new(@pub_keys))
+    rescue
+    else
+      assert(false)
+    end
+  end
```
Shouldn't those tests check that `Ssession` throws a particular error we expect? :thinking:

iamnotacake replied:
I'm not an expert in Ruby, my attempts to make it only catch `ThemisError` didn't succeed so far. BTW, the only possible exception we can catch is the `ThemisError`. However, it can contain different string error messages.

ilammy replied:
It looks something like this:

```ruby
  def test_invalid_key_public_ec
    assert_raise(Themis::ThemisError) do
      session = Themis::Ssession.new(
        'client', @client_pub_key, CallbacksForThemis.new(@pub_keys))
    end
  end
```

ilammy commented on tests/themispp/secure_session_test.hpp:
```diff
@@ -74,37 +75,7 @@ namespace secure_session_test
-                              0xf8, 0xf0, 0x66, 0x85, 0xee, 0xa3, 0x4c, 0x4f, 0xd5,
-                              0xb9, 0x85, 0xc2, 0xde, 0x39, 0x49, 0x8d, 0x99, 0xe1};
-static const std::vector<uint8_t> client_public_key = as_bytes(client_pub);
+/* Test keys can be found in secure_session_test_keys.hpp */
```
Thanks :heart:

ilammy commented:
@iamnotacake, currently I don't have my macBook available so I'm in the same boat as you: I can only make blind changes and test them by submitting it to CI :grin: Since you have been kind enough to enable the _Allow edits by maintainers_ check box, I have simply pushed the commit into this branch (please `git pull` it on your machine). Unfortunately, Secure Session is not tested at all in ObjCThemis so no tests at the moment.

ilammy commented:
Also, since this PR *does* change the behavior of some wrappers, I believe it should update the changelog for relevant languages.

iamnotacake commented:
> Also, since this PR _does_ change the behavior of some wrappers, I believe it should update the changelog for relevant languages.

Does this look good?
```
- **Python**

  - Fixed bug where `SSession` constructor did not throw an exception if `secure_session_create()` returned `NULL` ([#710](https://github.com/cossacklabs/themis/pull/710)).

- **Ruby**

  - Fixed bug where `Ssession` constructor did not throw an exception if `secure_session_create()` returned `NULL` ([#710](https://github.com/cossacklabs/themis/pull/710)).
```
BTW, should we put news in changelog in alphabetical order or append them (in the same `_Code:_` section) or it doesn't matter?
And if we add these things to changelog, I guess #698 should be mentioned too?

vixentael commented on tests/themispp/secure_session_test.hpp:
```diff
@@ -261,6 +232,42 @@ static void secure_session_ownership()
 }
 #endif
 
+static void secure_session_with_rsa_key()
```
```suggestion
static void secure_session_with_rsa_priv_key()
```

vixentael reviewed (commented):
Thank you! These are good changes.

I suggest to add test to check EC private keys as well (test should pass), because right now we have tests for RSA pub/priv and EC pub, which looks like EC priv are missing.

vixentael commented on tests/themispp/secure_session_test.hpp:
```diff
@@ -269,6 +276,9 @@ inline int run_secure_session_test()
     sput_run_test(secure_session_uninitialized, "secure_session_uninitialized", __FILE__);
     sput_run_test(secure_session_moved, "secure_session_moved", __FILE__);
     sput_run_test(secure_session_ownership, "secure_session_ownership", __FILE__);
+    sput_run_test(secure_session_with_rsa_key, "secure_session_with_rsa_key", __FILE__);
```
```suggestion
    sput_run_test(secure_session_with_rsa_priv_key, "secure_session_with_rsa_priv_key", __FILE__);
```

ilammy commented:
@vixentael,

> I suggest to add test to check EC private keys as well (test should pass)

Aren't EC private keys – the happy path – tested in other tests that verify that Secure Session works as expected? Or you suggest just a complementary test for the constructor alone to complete the puzzle?

vixentael commented:
> you suggest just a complementary test for the constructor alone to complete the puzzle?

yes, I suggest to complete the puzzle, because future us will totally forget why one case is not tested (which looks like we've missed that case)

ilammy commented:
@iamnotacake,

> Does this look good?
> [...]
> Maybe you mean mentioning that sec session wrappers don't allow RSA keys anymore? Or both this and bugfixes? Now I'm a lil bit confised

The changelog is mostly for the users and they probably don't (and shouldn't) know about Themis Core functions. So I'd write there something like "${Secure Session constructor name} now (throws an exception | returns `null` | does a dance) when given incorrect keys", or maybe RSA keys specifically.

The idea is to describe the expected behavior to the user. With this change, the constructors may start failing due to incorrect keys. This changelog entry might help debugging.

> BTW, should we put news in changelog in alphabetical order or append them (in the same _Code:_ section) or it doesn't matter?

I'd ask to keep them alphabetical. That's one predictable order to look for a language.

> And if we add these things to changelog, I guess #698 should be mentioned too?

Yeah, makes sense. Please do.

iamnotacake commented on src/wrappers/themis/jsthemis/test/test.js:
```diff
@@ -307,6 +308,22 @@ describe("jsthemis", function(){
                 expect_code(addon.INVALID_PARAMETER)
             )
         })
+        it('does not allow EC public key', function() {
```
I didn't add EC private since they are already involved in every Secure Session test that tries to encrypt and send data. Put :+1: if you think it's still better to add them (to all wrappers mentioned in this PR + jsthemis) or :-1: if I convinced you we don't need them.
Anyway, users are not like "oh, let's read some tests, here's private EC key and it doesn't raise exception so I can use it and here's private RSA and looks like I won't be able to use it", they read the docs instead.

vixentael replied:
1. That's not for users, that's for future us. Future us might forget that EC private key is a valid case (as we forgot about RSA keys in SS).
2. These are different types of tests.

While I agree that adding additional tests for EC private keys overlaps a bit with current test suite, I don't see any disadvantages of this approach, only improvements of maintainability.

Lagovas commented on src/wrappers/themis/python/pythemis/scomparator.py:
```diff
@@ -42,7 +42,7 @@ class SComparator(object):
         self.session_ctx = ctypes.POINTER(ctypes.c_int)
         self.comparator_ctx = scomparator_create()
-        if self.comparator_ctx is None:
+        if not self.comparator_ctx:
```
it's okay here. secure_comparator_create will return NULL (0 or None as python types) which will mean False in `if` condition

Lagovas commented on src/wrappers/themis/python/pythemis/ssession.py:
```diff
@@ -143,7 +143,7 @@ def __init__(self, user_id, sign_key, transport):
                 len(sign_key),
                 ctypes.byref(self.transport_))
-        if self.session_ctx is None:
+        if not self.session_ctx:
```
same

Lagovas commented on gothemis/session/session_test.go:
```diff
@@ -223,6 +223,50 @@ func testSession(keytype int, t *testing.T) {
 
+func TestValidKeyHandling(t *testing.T) {
+	kpa, err := keys.New(keys.TypeEC)
+	if nil != err {
```
common way to check errors is `if err != nil` in golang. It's okay do such checks in C to avoid errors like `if err = nil`. But in golang we cannot assign value in `if` statements

iamnotacake replied:
`if nil != err` was just a copypase, I would usually write `if err != nil` as well as `if a == 300` instead of `if 300 == a` which IMO looks kinda ugly.

BTW, not too long ago Python introduced `:=` operator just to use it in `if`s and loops

Lagovas replied:
yes, but gothemis in most cases will be read by golang developers, not python)

iamnotacake commented on gothemis/session/session_test.go:
```diff
@@ -223,6 +223,29 @@ func testSession(keytype int, t *testing.T) {
+
+	clb := &testCallbacks{kpa, kpb}
+
+	_, err = New(clientID, kpa.Private, clb)
```
You mean like this?
```diff
-    _, err = New(clientID, kpa.Private, clb)
-    if err != nil {
+    client, err = New(clientID, kpa.Private, clb)
+    if client == nil || err != nil {
```

Lagovas replied:
```
if err == nil || client != nil
```
We expect that err **should not** to be nil and client **should** be nil

ilammy commented on CHANGELOG.md:
```diff
@@ -12,22 +12,34 @@ _Code:_
   - `make deb` and `make rpm` with `ENGINE=boringssl` will now produce `libthemis-boringssl` packages with embedded BoringSSL ([#683](https://github.com/cossacklabs/themis/pull/683), [#686](https://github.com/cossacklabs/themis/pull/686)).
   - `secure_session_create()` now allows only EC keys, returning an error for RSA ([#693](https://github.com/cossacklabs/themis/pull/693)).
 
+- **Objective-C**
+
+  - Updated Objective-C examples (iOS and macOS, Carthage and CocoaPods) to showcase usage of the newest Secure Cell API: generating symmetric keys and using Secure Cell with Passphrase ([#688](https://github.com/cossacklabs/themis/pull/688)) and to use latest Themis 0.13.2 ([#701](https://github.com/cossacklabs/themis/pull/701), [#703](https://github.com/cossacklabs/themis/pull/703), [#706](http ...
```
Initializer of `TSSession` in Objective-C now also reports an error in case of invalid keys. It returns `nil` in case of error.

Ditto for Swift. The class is called the same, the result is the same.

vixentael commented on tests/themispp/secure_session_test_keys.hpp:
```diff
@@ -0,0 +1,157 @@
+namespace secure_session_test
+{
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatbility
```
```suggestion
// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatibility
```

vixentael commented on tests/themispp/secure_session_test_keys.hpp:
```diff
@@ -0,0 +1,157 @@
+                               0x02, 0x7e, 0x95, 0x7d, 0xcf, 0x6d, 0x9b, 0x4e, 0x4a};
+static const std::vector<uint8_t> server_private_key = as_bytes(server_priv);
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatbility
```
```suggestion
// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatibility
```

vixentael commented on tests/themispp/secure_session_test_keys.hpp:
```diff
@@ -0,0 +1,157 @@
+                              0x24, 0xeb, 0x37, 0xd3, 0xe7, 0xc7, 0xa2, 0xab, 0xdd};
+static const std::vector<uint8_t> server_public_key = as_bytes(server_pub);
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatbility
```
```suggestion
// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatibility
```

vixentael commented on tests/themispp/secure_session_test_keys.hpp:
```diff
@@ -0,0 +1,157 @@
+                               0x40, 0xf6, 0xea, 0xac, 0x8a, 0x75, 0x63, 0x67, 0xbc};
+static const std::vector<uint8_t> client_private_key = as_bytes(client_priv);
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatbility
```
```suggestion
// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatibility
```

vixentael commented on tests/themispp/secure_session_test_keys.hpp:
```diff
@@ -0,0 +1,157 @@
+                              0xb9, 0x85, 0xc2, 0xde, 0x39, 0x49, 0x8d, 0x99, 0xe1};
+static const std::vector<uint8_t> client_public_key = as_bytes(client_pub);
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatbility
```
```suggestion
// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatibility
```

vixentael commented on tests/themispp/secure_session_test_keys.hpp:
```diff
@@ -0,0 +1,157 @@
+};
+static const std::vector<uint8_t> client_private_rsa_key = as_bytes(client_priv_rsa);
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatbility
```
```suggestion
// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays): C++03 compatibility
```

ilammy reviewed (approved):
Yup. Nicely done, good to merge! :+1:
